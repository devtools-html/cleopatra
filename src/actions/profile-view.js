/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
