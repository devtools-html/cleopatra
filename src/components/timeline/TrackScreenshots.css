/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineTrackScreenshot {
  position: relative;
  background: var(--grey-20);
  overflow: hidden;
}

.timelineTrackScreenshotImgContainer {
  /* This positioning is defined through JavaScript */
  position: absolute;
  overflow: hidden;
  display: flex;
  justify-content: center;
}

.timelineTrackScreenshotImgContainer::after {
  /* Create a border on the right of the container */
  content: "";
  position: absolute;
  top: 0;
  width: 1px;
  height: 100%;
  background: var(--grey-30);
  right: 0;
}

.timelineTrackScreenshotHover {
  /* This positioning is defined through JavaScript */
  position: absolute;
  pointer-events: none;
  z-index: 4; /* Ensure this is higher than any other timeline element. */
}

.timelineTrackScreenshotHoverImg {
  border-radius: 5px;
  padding: 0.5px;
  border: 0.5px solid rgba(0,0,0,.2);
  box-shadow:0 2px 4px rgba(0,0,0,.2);
}
