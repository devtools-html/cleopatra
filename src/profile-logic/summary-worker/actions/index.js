/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { summarizeProfile } from '../../../profile-logic/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
