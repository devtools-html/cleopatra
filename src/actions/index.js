/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
