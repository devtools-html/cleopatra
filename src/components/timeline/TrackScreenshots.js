/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { createPortal } from 'react-dom';
import explicitConnect from '../../utils/connect';
import {
  getCommittedRange,
  getPreviewSelection,
} from '../../selectors/profile';
import { getThreadSelectors } from '../../selectors/per-thread';
import { withSize, type SizeProps } from '../shared/WithSize';

import type { ThreadIndex, Thread } from '../../types/profile';
import type { ScreenshotPayload } from '../../types/markers';
import type { TracingMarker } from '../../types/profile-derived';
import type { Milliseconds } from '../../types/units';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';

import { ensureExists } from '../../utils/flow';
import './TrackScreenshots.css';

type OwnProps = {|
  +threadIndex: ThreadIndex,
  +windowId: string,
  ...SizeProps,
|};
type StateProps = {|
  +thread: Thread,
  +rangeStart: Milliseconds,
  +rangeEnd: Milliseconds,
  +screenshots: TracingMarker[],
  +threadName: string,
  +isMakingPreviewSelection: boolean,
|};
type DispatchProps = {||};
type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
type State = {|
  offsetX: null | number,
  pageX: null | number,
  containerTop: null | number,
|};

// Export the value for tests.
export const TRACK_HEIGHT = 50;
const HOVER_HEIGHT = 100;
const HOVER_MAX_WIDTH_RATIO = 1.75;

class Screenshots extends PureComponent<Props, State> {
  state = {
    offsetX: null,
    pageX: null,
    containerTop: null,
  };

  _overlayElement = ensureExists(
    document.querySelector('#root-overlay'),
    'Expected to find a root overlay element.'
  );

  findScreenshotAtMouse(offsetX: number): number | null {
    const { width, rangeStart, rangeEnd, screenshots } = this.props;
    const rangeLength = rangeEnd - rangeStart;
    const mouseTime = offsetX / width * rangeLength + rangeStart;

    // Loop backwards to find the latest screenshot that has a time less
    // than the current time at the mouse position.
    for (let i = screenshots.length - 1; i >= 0; i--) {
      const screenshotTime = screenshots[i].start;
      if (mouseTime >= screenshotTime) {
        return i;
      }
    }

    return null;
  }

  /**
   * This function runs through all of the screenshots, and then samples the last known
   * screenshot, and places it on the screen, making a film strip.
   */
  renderScreenshotStrip() {
    const {
      thread,
      width: outerContainerWidth,
      rangeStart,
      rangeEnd,
      screenshots,
    } = this.props;

    if (screenshots.length === 0) {
      return null;
    }

    const images = [];
    const rangeLength = rangeEnd - rangeStart;
    const imageContainerWidth = TRACK_HEIGHT * 0.75;
    const timeToPixel = time =>
      outerContainerWidth * (time - rangeStart) / rangeLength;

    let screenshotIndex = 0;
    for (
      let left = timeToPixel(screenshots[0].start);
      left < outerContainerWidth;
      left += imageContainerWidth
    ) {
      // Try to find the next screenshot to fit in, or re-use the existing one.
      for (let i = screenshotIndex; i < screenshots.length; i++) {
        if (timeToPixel(screenshots[i].start) <= left) {
          screenshotIndex = i;
        } else {
          break;
        }
      }
      // Coerce the payload into a screenshot one.
      const payload: ScreenshotPayload = (screenshots[screenshotIndex]
        .data: any);
      const { url: urlStringIndex, windowWidth, windowHeight } = payload;
      const scaledImageWidth = TRACK_HEIGHT * windowWidth / windowHeight;
      images.push(
        <div
          className="timelineTrackScreenshotImgContainer"
          style={{ left, width: imageContainerWidth }}
          key={left}
        >
          {/* The following image is centered and cropped by the outer container. */}
          <img
            className="timelineTrackScreenshotImg"
            src={thread.stringTable.getString(urlStringIndex)}
            style={{
              width: scaledImageWidth,
              height: TRACK_HEIGHT,
            }}
          />
        </div>
      );
    }

    return images;
  }

  renderHoverPreview() {
    const { pageX, offsetX, containerTop } = this.state;
    const { screenshots, thread, isMakingPreviewSelection, width } = this.props;

    if (isMakingPreviewSelection || offsetX === null || pageX === null) {
      return null;
    }

    const screenshotIndex = this.findScreenshotAtMouse(offsetX);
    if (screenshotIndex === null) {
      return null;
    }

    // Coerce the payload into a screenshot one.
    const payload: ScreenshotPayload = (screenshots[screenshotIndex].data: any);
    const { url, windowWidth, windowHeight } = payload;

    // Compute the hover image's thumbnail size.
    let hoverHeight = HOVER_HEIGHT;
    let hoverWidth = HOVER_HEIGHT / windowHeight * windowWidth;

    if (hoverWidth > HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO) {
      // This is a really wide image, limit the height so it lays out reasonably.
      hoverWidth = HOVER_HEIGHT * HOVER_MAX_WIDTH_RATIO;
      hoverHeight = hoverWidth / windowWidth * windowHeight;
    }

    // Set the top so it centers around the track.
    const top = containerTop + (TRACK_HEIGHT - hoverHeight) * 0.5;
    const left =
      offsetX + hoverWidth * 0.5 > width
        ? // Stick the hover image on to the right side of the container.
          pageX - offsetX + width - hoverWidth
        : // Center the hover image around the mouse.
          pageX - hoverWidth * 0.5;

    return createPortal(
      <div className="timelineTrackScreenshotHover" style={{ left, top }}>
        <img
          className="timelineTrackScreenshotHoverImg"
          src={thread.stringTable.getString(url)}
          style={{
            height: hoverHeight,
            width: hoverWidth,
          }}
        />
      </div>,
      this._overlayElement
    );
  }

  _handleMouseLeave = () => {
    this.setState({
      offsetX: null,
      pageX: null,
      containerTop: null,
    });
  };

  _handleMouseMove = (event: SyntheticMouseEvent<HTMLDivElement>) => {
    const { top, left } = event.currentTarget.getBoundingClientRect();
    this.setState({
      pageX: event.pageX,
      offsetX: event.pageX - left,
      containerTop: top,
    });
  };

  render() {
    return (
      <div
        className="timelineTrackScreenshot"
        style={{ height: TRACK_HEIGHT }}
        onMouseLeave={this._handleMouseLeave}
        onMouseMove={this._handleMouseMove}
      >
        {this.renderScreenshotStrip()}
        {this.renderHoverPreview()}
      </div>
    );
  }
}

const options: ExplicitConnectOptions<OwnProps, StateProps, DispatchProps> = {
  mapStateToProps: (state, ownProps) => {
    const { threadIndex, windowId } = ownProps;
    const selectors = getThreadSelectors(threadIndex);
    const { start, end } = getCommittedRange(state);
    const previewSelection = getPreviewSelection(state);
    return {
      thread: selectors.getRangeFilteredThread(state),
      screenshots: ensureExists(
        selectors.getRangeFilteredScreenshotsById(state).get(windowId),
        'Expected to find screenshots for the given pid'
      ),
      threadName: selectors.getFriendlyThreadName(state),
      rangeStart: start,
      rangeEnd: end,
      isMakingPreviewSelection:
        previewSelection.hasSelection && previewSelection.isModifying,
    };
  },
  component: Screenshots,
};

export default withSize(explicitConnect(options));
