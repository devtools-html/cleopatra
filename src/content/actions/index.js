// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
