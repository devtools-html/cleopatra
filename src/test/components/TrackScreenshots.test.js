/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  Profile,
  Thread,
  IndexIntoMarkersTable,
} from '../../types/profile';

import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { commitRange } from '../../actions/profile-view';
import TrackScreenshots, {
  TRACK_HEIGHT,
} from '../../components/timeline/TrackScreenshots';
import Timeline from '../../components/timeline';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import mockRaf from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  getBoundingBox,
  addRootOverlayElement,
  removeRootOverlayElement,
} from '../fixtures/utils';
import { getScreenshotTrackProfile } from '../fixtures/profiles/make-profile';
import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';

// Mock out the getBoundingBox to have a 400 pixel width.
const TRACK_WIDTH = 400;
const LEFT = 5;
const TOP = 7;

describe('timeline/TrackScreenshots', function() {
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  it('matches the component snapshot', () => {
    const { view } = setup();
    expect(view).toMatchSnapshot();
    // Trigger any unmounting behavior handlers, just make sure it doesn't
    // throw any errors.
    view.unmount();
  });

  it('shows a hover when moving the mouse', () => {
    const { view } = setup();
    expect(view.find('.timelineTrackScreenshotHover').length).toBe(0);

    view.simulate('mousemove', { pageX: LEFT + 0 });
    view.update();
    expect(view.find('.timelineTrackScreenshotHover').length).toBe(1);
  });

  it('removes the hover when moving the mouse out', () => {
    const { view } = setup();
    expect(view.find('.timelineTrackScreenshotHover').length).toBe(0);

    view.simulate('mousemove', { pageX: LEFT + 0 });
    view.update();
    expect(view.find('.timelineTrackScreenshotHover').length).toBe(1);

    view.simulate('mouseleave');
    view.update();
    expect(view.find('.timelineTrackScreenshotHover').length).toBe(0);
  });

  it('moves the hover when moving the mouse', () => {
    const { moveMouseAndGetLeft } = setup();
    const base = moveMouseAndGetLeft(LEFT);
    expect(moveMouseAndGetLeft(LEFT + 10)).toBe(base + 10);
    expect(moveMouseAndGetLeft(LEFT + 20)).toBe(base + 20);
  });

  it('makes sure the hover image does not go off the end of the container', () => {
    const { moveMouseAndGetLeft } = setup();
    const pageX = LEFT + TRACK_WIDTH - 1;
    expect(pageX > moveMouseAndGetLeft(pageX)).toBe(true);
  });

  it('renders a screenshot images when zooming into a range without a screenshot start time actually in the range', () => {
    const profile = getScreenshotTrackProfile();
    const [thread] = profile.threads;
    const markerIndexA = thread.markers.length - 2;
    const markerIndexB = thread.markers.length - 1;

    _setScreenshotMarkersToUnknown(thread, markerIndexA, markerIndexB);

    const { dispatch, view } = setup(profile);
    dispatch(
      commitRange(
        thread.markers.time[markerIndexA],
        thread.markers.time[markerIndexB]
      )
    );
    view.update();
    expect(view.find('.timelineTrackScreenshotImg').length).toBeGreaterThan(0);
  });

  it('renders a no images when zooming into a range before screenshots', () => {
    const profile = getScreenshotTrackProfile();
    const [thread] = profile.threads;

    const markerIndexA = 0;
    const markerIndexB = 1;

    _setScreenshotMarkersToUnknown(thread, markerIndexA, markerIndexB);

    const { dispatch, view } = setup(profile);
    dispatch(
      commitRange(
        thread.markers.time[markerIndexA],
        thread.markers.time[markerIndexB]
      )
    );
    view.update();
    expect(view.find('.timelineTrackScreenshotImg').length).toBe(0);
  });

  it('is created in the <Timeline /> with a profile with screenshots', function() {
    const { view } = setup(getScreenshotTrackProfile(), <Timeline />);
    expect(view.find(TrackScreenshots).length).toBe(1);
  });

  it('is not created in the <Timeline /> with a profile with no screenshots', function() {
    const { view } = setup(getProfileWithNiceTracks(), <Timeline />);
    expect(view.find(TrackScreenshots).length).toBe(0);
  });
});

function setup(
  profile: Profile = getScreenshotTrackProfile(),
  component = <TrackScreenshots threadIndex={0} windowId="0" />
) {
  const store = storeWithProfile(profile);
  const { getState, dispatch } = store;
  const flushRafCalls = mockRaf();
  const ctx = mockCanvasContext();
  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => ctx);
  jest
    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
    .mockImplementation(() => {
      const rect = getBoundingBox(TRACK_WIDTH, TRACK_HEIGHT);
      // Add some arbitrary X offset.
      rect.left += LEFT;
      rect.right += LEFT;
      rect.x += LEFT;
      rect.y += TOP;
      rect.top += TOP;
      rect.bottom += TOP;
      return rect;
    });

  const view = mount(<Provider store={store}>{component}</Provider>);

  // WithSize uses requestAnimationFrame
  flushRafCalls();
  view.update();

  function moveMouseAndGetLeft(pageX: number): number {
    view.simulate('mousemove', { pageX });
    view.update();
    return view.find('.timelineTrackScreenshotHover').prop('style').left;
  }

  return {
    dispatch,
    getState,
    thread: profile.threads[0],
    store,
    view,
    moveMouseAndGetLeft,
  };
}

/**
 * Take a thread full screenshot markers, and set some to "Unknown" in order to
 * create gaps in a screenshot track.
 */
function _setScreenshotMarkersToUnknown(
  thread: Thread,
  ...markerIndexes: IndexIntoMarkersTable[]
) {
  // Remove off the last few screenshot markers
  const unknownStringIndex = thread.stringTable.indexForString('Unknown');
  const screenshotStringIndex = thread.stringTable.indexForString(
    'CompositorScreenshot'
  );
  for (const markerIndex of markerIndexes) {
    // Double check that we've actually got screenshot markers:
    if (thread.markers.name[markerIndex] !== screenshotStringIndex) {
      throw new Error('This is not a screenshot marker.');
    }
    thread.markers.name[markerIndex] = unknownStringIndex;
    thread.markers.data[markerIndex] = null;
  }
}
