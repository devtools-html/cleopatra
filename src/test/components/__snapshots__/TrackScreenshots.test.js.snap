// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/TrackScreenshots matches the component snapshot 1`] = `
<div
  className="timelineTrackScreenshot"
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  style={
    Object {
      "height": 50,
    }
  }
>
  Array [
    <div
      className="timelineTrackScreenshotImgContainer"
      key="0"
      style={
        Object {
          "left": 0,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="37.5"
      style={
        Object {
          "left": 37.5,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="75"
      style={
        Object {
          "left": 75,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="112.5"
      style={
        Object {
          "left": 112.5,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="150"
      style={
        Object {
          "left": 150,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="187.5"
      style={
        Object {
          "left": 187.5,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="225"
      style={
        Object {
          "left": 225,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="262.5"
      style={
        Object {
          "left": 262.5,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="300"
      style={
        Object {
          "left": 300,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="337.5"
      style={
        Object {
          "left": 337.5,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
    <div
      className="timelineTrackScreenshotImgContainer"
      key="375"
      style={
        Object {
          "left": 375,
          "width": 37.5,
        }
      }
    >
      <img
        className="timelineTrackScreenshotImg"
        src="CompositorScreenshot"
        style={
          Object {
            "height": 50,
            "width": 100,
          }
        }
      />
    </div>,
  ]
</div>
`;
